packages/flutter/lib/src/widgets/autocomplete.dart

```diff
@@ -0,0 +1,579 @@
+  /// The options are displayed floating below the field using a
+  /// [CompositedTransformFollower] inside of an [Overlay], not at the same
+  /// place in the widget tree as AutocompleteCore.
+  final AutocompleteOptionsViewBuilder<T> optionsViewBuilder;
+
+  /// Returns the string to display in the field when the option is selected.
+  ///
+  /// This is useful when using a custom T type and the string to display is
+  /// different than the string to search by.
+  ///
+  /// If not provided, will use `option.toString()`.
+  final AutocompleteOptionToString<T> displayStringForOption;
+
+  /// Called when an option is selected by the user.
+  ///
+  /// Any [TextEditingController] listeners will not be called when the user
+  /// selects an option, even though the field will update with the selected
+  /// value, so use this to be informed of selection.
+  final AutocompleteOnSelected<T>? onSelected;
+
+  /// A function that returns the current selectable options objects given the
+  /// current TextEditingValue.
+  final AutocompleteOptionsBuilder<T> optionsBuilder;
+
+  // The default way to convert an option to a string.
+  static String _defaultStringForOption(dynamic option) {
+    return option.toString();
+  }
+
+  @override
+  _AutocompleteCoreState<T> createState() => _AutocompleteCoreState<T>();
+}
+
+class _AutocompleteCoreState<T extends Object> extends State<AutocompleteCore<T>> {
+  final GlobalKey _fieldKey = GlobalKey();
+  final LayerLink _optionsLayerLink = LayerLink();
+  final TextEditingController _textEditingController = TextEditingController();
+  Iterable<T> _options = Iterable<T>.empty();
+  T? _selection;
+
+  // The OverlayEntry containing the options.
+  OverlayEntry? _floatingOptions;
+
+  // True iff the state indicates that the options should be visible.
+  bool get _shouldShowOptions {
+    final TextSelection selection = _textEditingController.selection;
+    final bool fieldIsFocused = selection.baseOffset >= 0
+        && selection.extentOffset >= 0;
+    final bool hasOptions = _options.isNotEmpty;
+    return fieldIsFocused && _selection == null && hasOptions;
+  }
+
+  // Called when _textEditingController changes.
+  void _onChangedField() {
+    final Iterable<T> options = widget.optionsBuilder(
+      _textEditingController.value,
+    );
+    assert(options != null);
+    setState(() {
+      _options = options;
+      if (_selection != null) {
+        final String selectionString = widget.displayStringForOption(_selection!);
+        if (_textEditingController.text == selectionString) {
+          return;
+        }
+      }
+      _selection = null;
+    });
+    _updateOverlay();
+  }
+
+  // Called from fieldViewBuilder when the user submits the field.
+  void _onFieldSubmitted() {
+    if (_options.isEmpty) {
+      return;
+    }
+    _select(_options.first);
+  }
+
+  // Select the given option and update the widget.
+  void _select(T? nextSelection) {
+    if (nextSelection == _selection) {
```

Comment:
if the user taps the same option then the tap is basically ignored (i.e. the options menu does not go away)?

Reply:
Also does this mean we want the developer to implement `==` and `hashCode`, if they're generating the list of options on the fly?

Reply:
Calling _select with null: Good point, they shouldn't.  I will make it non-nullable. If the user wants to clear the selection, they can just clear or change the field.

About comparison: Yes, I will document that.